extensions/amp-ad-network-base/0.1/amp-ad-network-base.js

```diff
@@ -0,0 +1,156 @@
+    this.initialSize_ = {
+      // TODO(levitzky) handle non-numeric values.
+      width: element.getAttribute('width'),
+      height: element.getAttribute('height'),
+    };
+
+    /** @const @private {!AmpAdContext} */
+    this.context_ = new AmpAdContext(this.win).setSize(this.initialSize_);
+  }
+
+  /**
+   * @param {./amp-ad-type-defs.ValidatorResultType} resultType
+   * @param {!./amp-ad-render.Renderer} renderer
+   * @final
+   */
+  bindRenderer(resultType, renderer) {
+    if (this.boundRenderers_[resultType]) {
+      dev().warn(TAG, `Rendering mode already bound for type '${resultType}'`);
+    }
+    this.boundRenderers_[resultType] = renderer;
+  }
+
+  /**
+   * @param {!./amp-ad-render.Validator} validator
+   * @final
+   */
+  bindValidator(validator) {
+    if (this.boundValidator_) {
+      dev().warn(TAG, 'Validator already bound.');
+    }
+    this.boundValidator_ = validator;
+  }
+
+  /**
+   * Collapses slot by setting its size to 0x0.
+   * @private
+   */
+  forceCollapse_() {
+    super.attemptChangeSize(0, 0);
+  }
+
+  /**
+   * @return {string} The finalized ad request URL.
+   * @protected
+   */
+  getRequestUrl() {
+    return '';
+  }
+
+  /**
+   * Processes the ad response as soon as the XHR request returns. This can be
+   * overridden and used as a hook to perform any desired logic before passing
+   * the response to the validator.
+   * @param {?../../../src/service/xhr-impl.FetchResponse} response
+   * @private
+   */
+  handleAdResponse_(response) {
+    if (!response.arrayBuffer) {
+      // TODO(levitzky) Add error reporting.
+      this.forceCollapse_();
+      return;
+    }
+    response.arrayBuffer().then(unvalidatedBytes => {
+      dev().assert(this.boundValidator_, 'Validator never bound!');
+      this.context_
+          .setUnvalidatedBytes(unvalidatedBytes)
+          .setHeaders(response.headers);
+      this.boundValidator_.validate(this.context_)
+          .then(context => this.handleValidatorResponse_(context))
+          .catch(error => this.handleValidatorError_(error));
+    });
+  }
+
+  /**
+   * Invoked whenever the ad response errors out for any reason whatsoever.
+   * @param {*} error
+   * @private
+   */
+  handleAdResponseError_(error) {
+    // TODO(levitzky) add actual error processing logic.
+    dev().warn(TAG, error);
+  }
+
+  /**
+   * Invoked whenever the validator encounters an error.
+   * @param {*} error
+   * @private
+   */
+  handleValidatorError_(error) {
+    // TODO(levitzky) add actual error processing logic.
+    dev().warn(TAG, error);
+  }
+
+  /**
+   * Processes validator response and delegates further action to appropriate
+   *   renderer.
+   * @param {!AmpAdContext} context
+   * @private
+   */
+  handleValidatorResponse_(context) {
+    const result = context.getValidatorResult();
+    dev().assert(this.boundRenderers_[result],
+        'Renderer for AMP creatives never bound!');
+    this.boundRenderers_[result].render(context, this);
```

Comment:
Can we free the renderer afterwards?

Reply:
Yes, but let's do it conditionally. If someone were to implement a version of this with refresh, they'd need to be able to keep the resources.